compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/UnsafeCompareAndSwapNode.java

```diff
@@ -98,4 +120,41 @@ public LocationIdentity getLocationIdentity() {
     public void lower(LoweringTool tool) {
         tool.getLowerer().lower(this, tool);
     }
+
+    @Override
+    public void virtualize(VirtualizerTool tool) {
+        ValueNode alias = tool.getAlias(object);
+        if (Options.VirtualCAS.getValue(tool.getOptions()) &&
+                        alias instanceof VirtualInstanceNode &&
+                        offset.isConstant()) {
+
+            VirtualInstanceNode obj = (VirtualInstanceNode) alias;
+
+            int index = resolveFieldIndex(obj);
+
+            if (index >= 0) {
+
+                ValueNode read = tool.getEntry(obj, index);
+
+                LogicNode equalsNode = CompareNode.createCompareNode(EQ, expected, read, tool.getConstantReflectionProvider(), NodeView.DEFAULT);
+
+                FixedGuardNode guardNode = new FixedGuardNode(equalsNode, DeoptimizationReason.Aliasing, DeoptimizationAction.InvalidateRecompile);
```

Comment:
A deopt without any alternative is a bit dangerous performance-wise: it might lead to deopt-loops. In particular here the "action" is `InvalidateRecompile` which is quite costly considering that the recompilation will lead to exactly the same code.

What about just using `ConditionalNodes`?
```java
ConditionalNode fieldValue = ConditionalNode.create(equalsNode, newValue, read);
ConditionalNode result = ConditionalNode.create(equalsNode);
//...
tool.setVirtualEntry(obj, index, fieldValue);
//...
tool.replaceWith(result);
```
In the "simple" cases it will immediately fold away and in more complex cases where equality can't be statically proven, performance should still be OK.